packages/block-library/src/heading/index.php

```diff
@@ -0,0 +1,93 @@
+<?php
+/**
+ * Server-side rendering of the `core/heading` block.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Adds a wp-block-heading class to the heading blocks without storing
+ * it in the serialized $content.
+ *
+ * @param array  $attributes Attributes of the block being rendered.
+ * @param string $content Content of the block being rendered.
+ *
+ * @return string The content of the block being rendered.
+ */
+function block_core_heading_render( $attributes, $content ) {
+	if ( ! $content ) {
+		return $content;
+	}
+	$matches = array();
+	$pattern = '/
+        ^(\s*)                               # Any leading whitespace.
+        <(?<tag_name>
+            h[1-6]                           # The opening tag...
+            (?=\s|>)                         # ...followed by a whitespace or >
+                                             # ?= means a "lookahead"
+        )
+        (?<before_class>                     # Any attributes prior to "class"
+            (?:                              # ?: is a "non-capturing group"
+                (?!\sclass=")[^>]            # Match all characters until ">" except when
+                                             # the next character sequence is \sclass="
+                                             # ?! is a "negative lookahead"
+            )*
+        )
+        (?:\s*class="(?<class_name>[^"]+)")? # The class attribute, if any
```

Comment:
I assumed the double quotes here since that's how Gutenberg serializes block markup. Happy to iterate on that if needed.

Reply:
if we're going to stop and acknowledge we're making this assumption why not avoid the need to make it?

```
class=(?P<quote>['"]).*?(\k{quote})
             |         |   \- until we find that quote character again
             |         \- un-greedy match of any character
             \- capture the quote character
```

Reply:
Good idea! I just updated this PR